File: LeetCode/494.Target_Sum/494.Target_Sum.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int sum = 0;
        for (int n : nums) {
            sum += n;
        }

        if (target > sum || target < -sum)
            return 0;
        if ((target + sum) % 2 != 0)
            return 0;

        // pos + neg = sum
        // pos - neg = target
        int pos = (sum + target) / 2;
        int[] dp = new int[pos + 1];
        dp[0] = 1;

        for (int i = 0; i < nums.length; i++) {
            for (int j = pos; j - nums[i] >= 0; j--) {
                dp[j] = dp[j - nums[i]] + dp[j];
            }
        }
        return dp[pos];
    }
}

File: LeetCode/494.Target_Sum/494.Target_Sum.swift
class Solution {
    func findTargetSumWays(_ nums: [Int], _ target: Int) -> Int {
        // sum(P) - sum(N) = target
        // sum(p) + sum(N) = sum(nums)
        // 2*sum(p) = target + sum(nums)

        let total = nums.reduce(0, +)
        if (total + target) % 2 == 1 { return 0 }
        let pTotal = (total + target) / 2

        if pTotal < 0 { return 0 }

        // how many way to add num to be pTotal
        let n = nums.count

        var dp = Array(repeating: 0, count: pTotal + 1)
        dp[0] = 1

        for num in nums {
            for j in stride(from: pTotal, through: num, by: -1) {
                dp[j] += dp[j - num]
            }
        }

        return dp[pTotal]
    }
}
